AppSyncRealTimeClient/ConnectionProvider/AppsyncRealtimeConnectionAsync/RealtimeConnectionProviderAsync.swift

```diff
@@ -0,0 +1,204 @@
+    /// alive" message will cause the timer to be reset to the full interval.
+    var staleConnectionTimer: CountdownTimer
+
+    /// Intermediate state when the connection is connected and connectivity updates to unsatisfied (offline)
+    var isStaleConnection: Bool
+
+    /// Manages concurrency for socket connections, disconnections, writes, and status reports.
+    ///
+    /// Each connection request will be sent to this queue. Connection request are
+    /// handled one at a time.
+    let taskQueue = TaskQueue<Void>()
+
+    /// Monitor for connectivity updates
+    let connectivityMonitor: ConnectivityMonitor
+
+    /// The serial queue on which status & message callbacks from the web socket are invoked.
+    private let serialCallbackQueue: DispatchQueue
+
+    /// Throttle when AppSync sends LimitExceeded error. High rate of subscriptions requests will cause AppSync to send
+    /// connection level LimitExceeded errors for each subscribe made. A connection level error means that there is no
+    /// subscription id associated with the error. When handling these errors, all subscriptions will receive a message
+    /// for the error. Use this subject to send and throttle the errors on the client side.
+    var limitExceededThrottleSink: Any?
+    var iLimitExceededSubject: Any?
+    var limitExceededSubject: PassthroughSubject<ConnectionProviderError, Never> {
+        if iLimitExceededSubject == nil {
+            iLimitExceededSubject = PassthroughSubject<ConnectionProviderError, Never>()
+        }
+        // swiftlint:disable:next force_cast
+        return iLimitExceededSubject as! PassthroughSubject<ConnectionProviderError, Never>
+    }
+
+    init(
+        url: URL,
+        websocket: AppSyncWebsocketProvider,
+
+        serialCallbackQueue: DispatchQueue = DispatchQueue(
+            label: "com.amazonaws.AppSyncRealTimeConnectionProvider.callbackQueue"
+        ),
+        connectivityMonitor: ConnectivityMonitor = ConnectivityMonitor()
+    ) {
+        self.url = url
+        self.websocket = websocket
+        self.listeners = [:]
+        self.status = .notConnected
+        self.staleConnectionTimer = CountdownTimer()
+        self.isStaleConnection = false
+        self.serialCallbackQueue = serialCallbackQueue
+        self.connectivityMonitor = connectivityMonitor
+
+        connectivityMonitor.start(onUpdates: handleConnectivityUpdates(connectivity:))
+    }
+
+    public convenience init(for url: URL, websocket: AppSyncWebsocketProvider) {
+        self.init(url: url, websocket: websocket)
+    }
+
+    // MARK: - ConnectionProvider methods
+
+    public func connect() {
+        taskQueue.async { [weak self] in
+            guard let self = self else {
+                return
+            }
+            guard self.status == .notConnected else {
+                self.updateCallback(event: .connection(self.status))
+                return
+            }
+            self.status = .inProgress
```

Comment:
Should this `status` state transition and the `updateCallback(event:)` occur after `websocket.connect(...)` is called? I'm mostly concerned about the client taking an action based on the `updateCallback(event:)` and the `interceptConnection` taking a while to complete, which could cause some funky state disconnect. 

If this concern is unnecessary, feel free to disregard.